File: Hash-Based Searching/Linear Hashing/LinearHashing.py
"""
    LinearHashing.py
    Python 3.x

    Features:
    - Dynamic Bucket Growth
    - Split Pointer (next_split)
    - Level-Based Hashing
    - Bucket Splitting
    - Overflow Buckets
    - Insert
    - Search
    - Display

    Run:
    python LinearHashing.py
"""

INITIAL_BUCKETS = 2
BUCKET_SIZE = 4


class Bucket:
    def __init__(self):
        self.keys = []
        self.overflow = None

    def is_full(self):
        return len(self.keys) >= BUCKET_SIZE


class LinearHashing:

    def __init__(self):

        self.level = 0
        self.next_split = 0

        self.buckets = [
            Bucket()
            for _ in range(INITIAL_BUCKETS)
        ]

    # ---------- Hash Functions ----------

    def hash_level(self, key):

        return (
            key %
            (INITIAL_BUCKETS << self.level)
        )

    def hash_next_level(self, key):

        return (
            key %
            (
                INITIAL_BUCKETS
                << (self.level + 1)
            )
        )

    def bucket_index(self, key):

        idx = self.hash_level(key)

        if idx < self.next_split:
            idx = self.hash_next_level(key)

        return idx

    # ---------- Insert Into Bucket ----------

    def insert_into_bucket(
        self,
        bucket,
        key
    ):

        current = bucket

        while True:

            if not current.is_full():

                current.keys.append(key)
                return

            if current.overflow is None:
                current.overflow = Bucket()

            current = current.overflow

    # ---------- Split Bucket ----------

    def split_bucket(self):

        split_index = self.next_split

        self.buckets.append(Bucket())

        values = []

        current = self.buckets[split_index]

        while current:

            values.extend(current.keys)

            current = current.overflow

        self.buckets[split_index] = Bucket()

        for value in values:

            idx = self.hash_next_level(
                value
            )

            self.insert_into_bucket(
                self.buckets[idx],
                value
            )

        self.next_split += 1

        round_size = (
            INITIAL_BUCKETS
            << self.level
        )

        if self.next_split == round_size:

            self.next_split = 0
            self.level += 1

    # ---------- Insert ----------

    def insert(self, key):

        idx = self.bucket_index(key)

        self.insert_into_bucket(
            self.buckets[idx],
            key
        )

        self.split_bucket()

    # ---------- Search ----------

    def search(self, key):

        idx = self.bucket_index(key)

        current = self.buckets[idx]

        while current:

            if key in current.keys:
                return True

            current = current.overflow

        return False

    # ---------- Display ----------

    def display(self):

        print(
            f"\nLevel: {self.level}"
        )

        print(
            f"Next Split: "
            f"{self.next_split}"
        )

        print(
            f"Bucket Count: "
            f"{len(self.buckets)}\n"
        )

        for i, bucket in enumerate(
            self.buckets
        ):

            print(
                f"Bucket[{i}] : ",
                end=""
            )

            current = bucket

            while current:

                for value in current.keys:
                    print(
                        value,
                        end=" "
                    )

                if current.overflow:
                    print(
                        "-> ",
                        end=""
                    )

                current = current.overflow

            print()


# ---------- Driver ----------

if __name__ == "__main__":

    hash_table = LinearHashing()

    values = [
        5, 12, 25, 33,
        41, 18, 29, 37,
        49, 52, 61, 73
    ]

    for value in values:
        hash_table.insert(value)

    hash_table.display()

    print(
        "\nSearch 29:",
        "Found"
        if hash_table.search(29)
        else "Not Found"
    )

    print(
        "Search 100:",
        "Found"
        if hash_table.search(100)
        else "Not Found"
    )

File: Hash-Based Searching/Linear Hashing/LinearHashing.js
/*
    LinearHashing.js
    ES2020+

    Features:
    - Dynamic Bucket Growth
    - Split Pointer (nextSplit)
    - Level-Based Hashing
    - Bucket Splitting
    - Overflow Buckets
    - Insert
    - Search
    - Display

    Run:
    node LinearHashing.js
*/

const INITIAL_BUCKETS = 2;
const BUCKET_SIZE = 4;

/* ---------- Bucket ---------- */

class Bucket {
    constructor() {
        this.keys = [];
        this.overflow = null;
    }

    isFull() {
        return this.keys.length >= BUCKET_SIZE;
    }
}

/* ---------- Linear Hashing ---------- */

class LinearHashing {

    constructor() {

        this.level = 0;
        this.nextSplit = 0;

        this.buckets = [];

        for (
            let i = 0;
            i < INITIAL_BUCKETS;
            i++
        ) {
            this.buckets.push(
                new Bucket()
            );
        }
    }

    /* ---------- Hash Functions ---------- */

    hashLevel(key) {

        return (
            key %
            (INITIAL_BUCKETS << this.level)
        );
    }

    hashNextLevel(key) {

        return (
            key %
            (
                INITIAL_BUCKETS
                << (this.level + 1)
            )
        );
    }

    bucketIndex(key) {

        let idx =
            this.hashLevel(key);

        if (idx < this.nextSplit) {

            idx =
                this.hashNextLevel(key);
        }

        return idx;
    }

    /* ---------- Insert Into Bucket ---------- */

    insertIntoBucket(bucket, key) {

        let current = bucket;

        while (true) {

            if (!current.isFull()) {

                current.keys.push(key);
                return;
            }

            if (
                current.overflow === null
            ) {
                current.overflow =
                    new Bucket();
            }

            current =
                current.overflow;
        }
    }

    /* ---------- Split Bucket ---------- */

    splitBucket() {

        const splitIndex =
            this.nextSplit;

        this.buckets.push(
            new Bucket()
        );

        const values = [];

        let current =
            this.buckets[splitIndex];

        while (current) {

            values.push(
                ...current.keys
            );

            current =
                current.overflow;
        }

        this.buckets[
            splitIndex
        ] = new Bucket();

        for (const value of values) {

            const idx =
                this.hashNextLevel(
                    value
                );

            this.insertIntoBucket(
                this.buckets[idx],
                value
            );
        }

        this.nextSplit++;

        const roundSize =
            INITIAL_BUCKETS
            << this.level;

        if (
            this.nextSplit ===
            roundSize
        ) {

            this.nextSplit = 0;
            this.level++;
        }
    }

    /* ---------- Insert ---------- */

    insert(key) {

        const idx =
            this.bucketIndex(key);

        this.insertIntoBucket(
            this.buckets[idx],
            key
        );

        this.splitBucket();
    }

    /* ---------- Search ---------- */

    search(key) {

        const idx =
            this.bucketIndex(key);

        let current =
            this.buckets[idx];

        while (current) {

            if (
                current.keys.includes(
                    key
                )
            ) {
                return true;
            }

            current =
                current.overflow;
        }

        return false;
    }

    /* ---------- Display ---------- */

    display() {

        console.log(
            `\nLevel: ${this.level}`
        );

        console.log(
            `Next Split: ${this.nextSplit}`
        );

        console.log(
            `Bucket Count: ${this.buckets.length}\n`
        );

        for (
            let i = 0;
            i < this.buckets.length;
            i++
        ) {

            let output =
                `Bucket[${i}] : `;

            let current =
                this.buckets[i];

            while (current) {

                for (
                    const value
                    of current.keys
                ) {
                    output +=
                        value + " ";
                }

                if (
                    current.overflow
                ) {
                    output += "-> ";
                }

                current =
                    current.overflow;
            }

            console.log(output);
        }
    }
}

/* ---------- Driver ---------- */

const hashTable =
    new LinearHashing();

const values = [
    5, 12, 25, 33,
    41, 18, 29, 37,
    49, 52, 61, 73
];

for (const value of values) {
    hashTable.insert(value);
}

hashTable.display();

console.log(
    "\nSearch 29:",
    hashTable.search(29)
        ? "Found"
        : "Not Found"
);

console.log(
    "Search 100:",
    hashTable.search(100)
        ? "Found"
        : "Not Found"
);
